zproject/prod_settings_template.py

```diff
@@ -1,5 +1,7 @@
 from typing import Any, Dict, Tuple
 
+from zerver.lib.types import OIDCIdPConfigDict, SAMLIdPConfigDict
```

Comment:
`prod_settings_template.py` is copied to `/etc/zulip/settings.py` on installation and not automatically updated after that. So this would force us to maintain these members of `zerver.lib.types` forever (or write some kind of migration). Guarding this with `if TYPE_CHECKING` would be an easy way to avoid imposing that compatibility constraint on ourselves.

Reply:
Is it the same reason for us to do the same thing in `zproject.default_settings`? If I understand correctly, we don't need to conditionally import the types in `zproject.test_extra_settings` and `zproject.computed_settings`, right?

Reply:
Correct, `prod_settings_template` is the only file that gets copied in this way. (A possible reason to conditionally import in those other files would be to avoid import cycles, but `zerver.lib.types` wouldn’t cycle.)

Reply:
This still feels risky to me. I suppose it's not obviously harmful to import something that no longer exists inside an `if False`. Do we have other reasonable options for how to handle this?

Reply:
Well, it would be way more reasonable to use a TOML or INI configuration file. Letting users execute code here, and clobbering the Django settings with whatever variables they create, is kind of asking for compatibility trouble; I’m guessing the only reason we don’t see it in practice is that users aren’t writing anything other than plain data.

However, if we’re going to use typed Python here, `if TYPE_CHECKING` is exactly the right mechanism for avoiding the compatibility hazard. If something breaks in this system, I don’t think it’s going to be that.

Reply:
Yeah, I guess this is OK -- there's no code change that actually runs outside type-checking. It's not going to be a problem if types don't check on someone's production system.

As a sidenote: We're stuck with a Python file for configuration of at least `django-auth-ldap`, which is a library that we might want to think about replacing or substantially forking anyway, since it's configuration structure isn't compatible with subdomains anyway. Could be a useful #backend conversation, though we may have higher priorities.

Reply:
If the concerns remain, how about we fall back to `Dict[str, Any]` only in `prod_settings_template` to make it type check? If we are not planning to keep this implementation, it might be better off to just avoid dependencies and opt-out from fully type-checking this particular configuration for now.

Reply:
Possibly we can roll out support for a different configuration format and slowly deprecate the old format in the future. Which should be pretty beneficial for tooling and managing compatibility once it is complete.

Reply:
> We're stuck with a Python file for configuration of at least `django-auth-ldap`.

This is entirely a self-inflicted problem. Part of what I’m saying is that we should not be exposing the settings interfaces of our dependencies directly to users.